File: medium/0443_string_compression/solution.cpp
#include <string>
#include <vector>

class Solution {
public:
  int compress(std::vector<char> &chars) {
    const size_t n = chars.size();

    int compressedCharsLen = 0;

    for (size_t right{}; right < n; ++right) {
      size_t left = right;
      while (right + 1 < n && chars[right] == chars[right + 1])
        ++right;

      chars[compressedCharsLen++] = chars[left];

      if (right - left + 1 > 1) {
        for (const char &digit : std::to_string(right - left + 1))
          chars[compressedCharsLen++] = digit;
      }
    }

    return compressedCharsLen;
  }
};

File: medium/0443_string_compression/solution.py
from typing import List


class Solution:
    def compress(self, chars: List[str]) -> int:
        n = len(chars)

        read = 0
        write = 0

        while read < n:
            chars[write] = chars[read]
            write += 1
            read += 1

            count = 1

            while read < n and chars[read - 1] == chars[read]:
                read += 1
                count += 1

            if count > 1:
                for digit in str(count):
                    chars[write] = digit
                    write += 1

        return write
